contracts/nftauction.js

```diff
@@ -0,0 +1,597 @@
+      params.creationFee = creationFee;
+    }
+    if (minBidIncrementPercent) {
+      if (!api.assert(Number.isInteger(minBidIncrementPercent) && minBidIncrementPercent > 0 && minBidIncrementPercent <= MAX_BID_INCREMENT_PERCENT, 'invalid minBidIncrementPercent')) return;
+      params.minBidIncrementPercent = minBidIncrementPercent;
+    }
+    if (cancelLockTimeMillis) {
+      if (!api.assert(Number.isInteger(cancelLockTimeMillis) && cancelLockTimeMillis > 0, 'invalid cancelLockTimeMillis')) return;
+      params.cancelLockTimeMillis = cancelLockTimeMillis;
+    }
+    if (expiryTimeMillis) {
+      if (!api.assert(Number.isInteger(expiryTimeMillis) && expiryTimeMillis > 0, 'invalid expiryTimeMillis')) return;
+      params.expiryTimeMillis = expiryTimeMillis;
+    }
+    if (maxExpiryTimeMillis) {
+      if (!api.assert(Number.isInteger(maxExpiryTimeMillis) && maxExpiryTimeMillis > 0, 'invalid maxExpiryTimeMillis')) return;
+      params.maxExpiryTimeMillis = maxExpiryTimeMillis;
+    }
+    if (auctionsPerBlock) {
+      if (!api.assert(Number.isInteger(auctionsPerBlock) && auctionsPerBlock > 0, 'invalid auctionsPerBlock')) return;
+      params.auctionsPerBlock = auctionsPerBlock;
+    }
+
+    await api.db.update('params', params);
+  }
+};
+
+const countDecimals = value => api.BigNumber(value).dp();
+
+const getTimestamp = (value) => {
+  try {
+    const date = new Date(`${value}.00Z`);
+    return date.getTime();
+  } catch (e) {
+    return false;
+  }
+};
+
+const transferIsSuccessful = (result, action, from, to, symbol, quantity) => {
+  if (result.errors === undefined
+    && result.events && result.events.find(el => el.contract === 'tokens'
+    && el.event === action
+    && el.data.from === from
+    && el.data.to === to
+    && api.BigNumber(el.data.quantity).eq(quantity)
+    && el.data.symbol === symbol) !== undefined) {
+    return true;
+  }
+
+  return false;
+};
+
+const sendNfts = async (to, wrappedNfts) => {
+  // transfer NFTs
+  await api.executeSmartContract('nft', 'transfer', {
+    fromType: 'contract',
+    to,
+    toType: 'user',
+    nfts: [wrappedNfts],
+    isSignedWithActiveKey: true,
+  });
+};
+
+const returnBids = async (bids, priceSymbol) => {
+  for (let i = 0; i < bids.length; i += 1) {
+    const {
+      account,
+      bid,
+    } = bids[i];
+
+    // send tokens back to bidders
+    await api.transferTokens(account, priceSymbol, bid, 'user');
+  }
+};
+
+const cancelAuction = async (auction) => {
+  const {
+    auctionId,
+    symbol,
+    seller,
+    nftIds,
+    priceSymbol,
+    bids,
+  } = auction;
+
+  // return all the bids
+  await returnBids(bids, priceSymbol);
+
+  const wrappedNfts = {
+    symbol,
+    ids: nftIds,
+  };
+
+  // return NFTs to the seller
+  await sendNfts(seller, wrappedNfts);
+
+  api.emit('cancelAuction', {
+    auctionId,
+  });
+};
+
+const settleAuction = async (auction, id = null) => {
```

Comment:
can we name 'id' here as 'index', so it's clear it's not some numeric identifier on the bid